### PR TITLE
[FLINK-15442] Harden the Avro Confluent Schema Registry nightly end-to-end test

### PR DESCRIPTION
**Pull request by KarmaGYZ**
## What is the purpose of the change

Harden the Avro Confluent Schema Registry nightly end-to-end test.

## Brief change log

- Wrap the error when execute cleanup command in `retry_times_with_backoff_and_cleanup` function.
- Harden the way to loop in `retry_times_with_backoff_and_cleanup`

## Verifying this change

Trigger the Avro Confluent Schema Registry nightly end-to-end test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by KarmaGYZ** [comment]
cc @tillrohrmann 
Travis gives green light to the relevant test.
https://travis-ci.org/KarmaGYZ/flink/builds/632126284

File: flink-end-to-end-tests/test-scripts/common.sh [hunk 1]
```diff
@@ -748,7 +748,7 @@ function retry_times_with_backoff_and_cleanup() {
     local command="$3"
     local cleanup_command="$4"
 
-    for (( i = 0; i < ${retriesNumber}; i++ ))
+    for i in $(seq 1 ${retriesNumber})
```
Comment (tillrohrmann) [inline flink-end-to-end-tests/test-scripts/common.sh]: Why is this better than the previous for loop version?

File: flink-end-to-end-tests/test-scripts/kafka-common.sh [hunk 2]
```diff
@@ -36,7 +36,7 @@ function setup_kafka_dist {
   mkdir -p $TEST_DATA_DIR
   KAFKA_URL="https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.11-$KAFKA_VERSION.tgz"
   echo "Downloading Kafka from $KAFKA_URL"
-  curl "$KAFKA_URL" --retry 10 --retry-max-time 120 > $TEST_DATA_DIR/kafka.tgz
+  retry_times 5 0 "curl ${KAFKA_URL} --retry 10 --retry-max-time 120 --output ${TEST_DATA_DIR}/kafka.tgz"
```
Comment (tillrohrmann) [inline flink-end-to-end-tests/test-scripts/kafka-common.sh]: Weren't we already retrying the curl command by setting `--retry 10`?

File: flink-end-to-end-tests/test-scripts/common.sh (hunk 1 above)
Comment (KarmaGYZ) [reply flink-end-to-end-tests/test-scripts/common.sh]: This block of loop will only executes 1 time in this scenario. I found that there is also a for loop  in start_confluent_schema_registry function:
```
for i in {1..30}; do
    if get_and_verify_schema_subjects_exist; then
        echo "Schema registry is up."
        return 0
    fi
    echo "Waiting for schema registry..."
    sleep 1
  done
```
The variable i  is set to 30 after this function, which cause the loop end after only 1 execute.

File: flink-end-to-end-tests/test-scripts/kafka-common.sh (hunk 2 above)
Comment (KarmaGYZ) [reply flink-end-to-end-tests/test-scripts/kafka-common.sh]: Thanks for mention this. Yes, we could rely on the curl native mechanism. However, after reading the [curl manual](https://curl.haxx.se/docs/manpage.html). I found it's better to add the `--retry-delay` configure here to disable the exponential backoff mechanism. WDYT?

**Comment by KarmaGYZ** [comment]
Thanks for the review @tillrohrmann . I've updated the PR.

File: flink-end-to-end-tests/test-scripts/kafka-common.sh (hunk 2 above)
Comment (tillrohrmann) [reply flink-end-to-end-tests/test-scripts/kafka-common.sh]: En exponential backoff is not necessarily bad. We've capped it to 120s which sounds fair to me. Hence, I don't think that we have to change it.

File: flink-end-to-end-tests/test-scripts/common.sh (hunk 1 above)
Comment (tillrohrmann) [reply flink-end-to-end-tests/test-scripts/common.sh]: I tried to reproduce the problem via

```
for i in {1..30}; do
  echo $i
done

for (( i = 0; i < 10; i++ )) do
  echo $i
done
```

When I execute this, then 1..30 and 0..9 will be printed. Hence, I'm not sure I really understand the problem you are describing. Could you provide me an example script to reproduce the problem with the old for loop?

File: flink-end-to-end-tests/test-scripts/kafka-common.sh [hunk 3]
```diff
@@ -36,7 +36,7 @@ function setup_kafka_dist {
   mkdir -p $TEST_DATA_DIR
   KAFKA_URL="https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.11-$KAFKA_VERSION.tgz"
   echo "Downloading Kafka from $KAFKA_URL"
-  curl "$KAFKA_URL" --retry 10 --retry-max-time 120 > $TEST_DATA_DIR/kafka.tgz
+  curl ${KAFKA_URL} --retry 10 --retry-delay 2 --retry-max-time 120 --output ${TEST_DATA_DIR}/kafka.tgz
```
Comment (tillrohrmann) [inline flink-end-to-end-tests/test-scripts/kafka-common.sh]: I think an exponential backoff strategy is usually something good. Hence, I would suggest to remove `--retry-delay`.

File: flink-end-to-end-tests/test-scripts/kafka-common.sh [hunk 4]
```diff
@@ -50,7 +50,7 @@ function setup_confluent_dist {
   mkdir -p $TEST_DATA_DIR
   CONFLUENT_URL="http://packages.confluent.io/archive/$CONFLUENT_MAJOR_VERSION/confluent-oss-$CONFLUENT_VERSION-2.11.tar.gz"
   echo "Downloading confluent from $CONFLUENT_URL"
-  curl "$CONFLUENT_URL" > $TEST_DATA_DIR/confluent.tgz
+  curl ${CONFLUENT_URL} --retry 10 --retry-delay 2 --retry-max-time 120 --output ${TEST_DATA_DIR}/confluent.tgz
```
Comment (tillrohrmann) [inline flink-end-to-end-tests/test-scripts/kafka-common.sh]: Same here with `retry-delay`.

File: flink-end-to-end-tests/test-scripts/common.sh (hunk 1 above)
Comment (KarmaGYZ) [reply flink-end-to-end-tests/test-scripts/common.sh]: I think current scenario equals to the following script:
```
for (( i = 0; i < 10; i++ )) do
  for i in {1..30}; do
     echo $i
  done
  echo $i
done
```

**Comment by KarmaGYZ** [comment]
Thanks for the review @tillrohrmann . I've removed the `retry-delay` and explained what happened in the loop.

File: flink-end-to-end-tests/test-scripts/common.sh (hunk 1 above)
Comment (tillrohrmann) [reply flink-end-to-end-tests/test-scripts/common.sh]: Ah this makes sense. Thanks a lot for your explanation.

**Pull request closed without merging**